lib/Doctrine/Enumeration/Enumerated.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Enumeration;
+
+use Doctrine\Enumeration\Exception\NotSupportedException;
+use function array_search;
+use function constant;
+use function in_array;
+
+trait Enumerated
+{
+    final private function __construct()
+    {
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function validateValue($value) : void
+    {
+        if (! in_array($value, self::availableValues(), true)) {
+            throw new NotSupportedException('Invalid value.');
+        }
+    }
+
+    final public static function validateName(string $name) : void
+    {
+        if (! isset(self::availableValues()[$name])) {
+            throw new NotSupportedException('Invalid name.');
+        }
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function get($value) : self
+    {
+        self::validateValue($value);
+
+        $name      = array_search($value, self::availableValues(), true);
+        $instances = &self::registry(static::class);
+
+        return $instances[$name] ?? $instances[$name] = new static();
+    }
+
+    /**
+     * @return mixed
+     */
+    final private static function value(self $instance)
+    {
+        return self::availableValues()[array_search($instance, self::registry(static::class), true)];
```

Comment:
Maybe instead of `array_search()` you could use either `spl_object_hash()` or `SplObjectStorage()`. However, I'm not sure if it'd be more efficient given that usually, the array will contain a handful of values.